A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,440 @@
+value is `/Package1.Service2` and the RPC method is
+`/Package2.Service1/Method3` then just pass the RPC through. Note, for path
+matching the rules described in [rfc6265 section-5.1.4][rfc6265-path-match]
+will apply.
+
+* Search the RPC headers (metadata) for header(s) named `Cookie` and get the
+set of all the cookies present. If the set does not contain a cookie whose
+name matches the `name` configuration value of the filter, then pass the RPC
+through. If you find more than one cookie with that `name` then log a local
+warning and pass the RPC through. Otherwise get the value of that cookie and
+base64-decode the value to get the `override-host` for the RPC. This value
+should be a syntactically valid IP:port address: if not, then log a local
+warning and pass the RPC through. For a valid `override-host` value, set it
+in the RPC context as the value of `upstreamAddress` state variable which
+will be used in the response processing of that RPC.
+
+* In case of a valid `override-host` value, pass this value as a
+"pick-argument" to the Load Balancer's pick method as described in [A31][].
+For example,
+
+    * in Java use `CallOptions` to pass this value.
+
+    * in Go pass it via `Context`.
+
+    * In C-core, this will be via the `GRPC_CONTEXT_SERVICE_CONFIG_CALL_DATA`
+      call context element whose API will be extended to allow the
+      filter to add a new call element.
+
+* In the response path - if the filter was not skipped based on the RPC
+  path - get the value of `upstreamAddress`. Also get the peer address for the
+  RPC; let’s call it `hostAddress` here. For example, in Java you get the peer
+  address through the `Grpc.TRANSPORT_ATTR_REMOTE_ADDR` attribute of the
+  current ClientCall. If `upstreamAddress` is not set or `upstreamAddress` and
+  `hostAddress` are different, then create a cookie using our Cookie config
+  which has the `name`, `ttl` and `path` values. Set the value of the cookie
+  to be the base64-encoded string value of `hostAddress`. Add this Cookie to
+  the response headers. The logic in pseudo-code is:
+
+```C++
+  if (!upstreamAddress.has_value() || hostAddress != upstreamAddress) {
+    String encoded_address = base64_encode(hostAddress);
+    Create a Cookie from our cookie configuration;
+    Set cookie's value as encoded_address;
+    Add the cookie header to the response;
+  }
+```
+
+### LB Policy for Stateful Session Affinity
+
+After the `CookieBasedStatefulSessionFilter` has passed the `override-host`
+value to the Load Balancer as a "pick-argument", the Load Balancer uses
+this value to route the RPC appropriately. The required logic needs to be
+implemented in a new and separate LB policy at an appropriate place in the
+hierarchy.
+
+RPC load reporting happens in the `xds_cluster_impl_experimental` policy and
+we do want all RPCs to be included in the load reports. Hence the new policy
+needs to be just below the `xds_cluster_impl_experimental` as its child policy.
```

Comment:
We need to call out that this means that the `override_host` policy will be beneath the `priority` policy, which means that if we pick a host for a session while we are failed over to a lower priority and then a higher priority comes back online, we will switch hosts at that point, which is different from Envoy.  I think this will be a fairly rare situation, so hopefully it won't be a big problem.  But it's not something we can easily change, because load reporting is done at a lower level of the tree for us than it is for Envoy.

Reply:
done

Reply:
> We need to call out that this means that the `override_host` policy will be beneath the `priority` policy, which means that if we pick a host for a session while we are failed over to a lower priority and then a higher priority comes back online, we will switch hosts at that point, which is different from Envoy. I think this will be a fairly rare situation, so hopefully it won't be a big problem.

When and how does this priority failover happen?

Reply:
@sanjaypujare This was very helpful. Can you point to an xDS/Istio based use-case which would translate to creation of `priority_experimental` policy

Reply:
Locality failover https://istio.io/latest/docs/tasks/traffic-management/locality-load-balancing/failover/ which says

"Internally, [Envoy priorities](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/priority.html) are used to control failover."